### ISSUE TITLE
Java application can't start with the current master

### ISSUE THREAD

**Issue by giner** [cloudfoundry/java-buildpack#392]
Some applications silently fails to start with a914fd6 release of the build pack without providing any meaningful information.

Bellow I'm trying to deploy the same application with two different versions of the buildpack (a914fd6 and v3.14) and three different settings for memory limit.

# v3.14 behavior
```
Starting app v3.14-128M
-----> Downloaded app package (32M)
Cloning into '/tmp/buildpacks/java-buildpack'...
Note: checking out '26cf03bdf49c93f01793d0f2753e41a3e7061bc9'.
You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.
If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:
  git checkout -b new_branch_name
-----> Java Buildpack Version: v3.14 | https://github.com/rakutentech/java-buildpack#26cf03b
-----> Downloading Open Jdk JRE 1.8.0_121 from https://java-buildpack.cloudfoundry.org/openjdk/trusty/x86_64/openjdk-1.8.0_121.tar.gz (0.5s)
       Expanding Open Jdk JRE to .java-buildpack/open_jdk_jre (1.1s)
-----> Downloading Open JDK Like Memory Calculator 2.0.2_RELEASE from https://java-buildpack.cloudfoundry.org/memory-calculator/trusty/x86_64/memory-calculator-2.0.2_RELEASE.tar.gz (0.0s)
       Cannot balance memory: Memory allocation failed for configuration: [metaspace:64M.. stack:228K..], : memory exceeded
[Buildpack]                      ERROR Compile failed with exception RuntimeError
Staging failed: Buildpack compilation step failed

FAILED
Error restarting application: BuildpackCompileFailed

TIP: use 'cf logs v3.14-128M --recent' for more information
```
```
Starting app v3.14-256M
-----> Downloaded app package (32M)
Cloning into '/tmp/buildpacks/java-buildpack'...
Note: checking out '26cf03bdf49c93f01793d0f2753e41a3e7061bc9'.
You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.
If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:
  git checkout -b new_branch_name
-----> Java Buildpack Version: v3.14 | https://github.com/rakutentech/java-buildpack#26cf03b
-----> Downloading Open Jdk JRE 1.8.0_121 from https://java-buildpack.cloudfoundry.org/openjdk/trusty/x86_64/openjdk-1.8.0_121.tar.gz (0.5s)
       Expanding Open Jdk JRE to .java-buildpack/open_jdk_jre (1.1s)
-----> Downloading Open JDK Like Memory Calculator 2.0.2_RELEASE from https://java-buildpack.cloudfoundry.org/memory-calculator/trusty/x86_64/memory-calculator-2.0.2_RELEASE.tar.gz (0.1s)
       Memory Settings: -Xss228K -Xmx104169K -XX:MaxMetaspaceSize=64M -Xms104169K -XX:MetaspaceSize=64M
-----> Downloading Container Certificate Trust Store 2.0.0_RELEASE from https://java-buildpack.cloudfoundry.org/container-certificate-trust-store/container-certificate-trust-store-2.0.0_RELEASE.jar (0.0s)
       Adding certificates to .java-buildpack/container_certificate_trust_store/truststore.jks (0.3s)
-----> Downloading Spring Auto Reconfiguration 1.10.0_RELEASE from https://java-buildpack.cloudfoundry.org/auto-reconfiguration/auto-reconfiguration-1.10.0_RELEASE.jar (0.1s)
-----> Uploading droplet (78M)

0 of 1 instances running, 1 starting
0 of 1 instances running, 1 starting
0 of 1 instances running, 1 starting
0 of 1 instances running, 1 down
0 of 1 instances running, 1 down
0 of 1 instances running, 1 down
```
```
Starting app v3.14-512M
-----> Downloaded app package (32M)
Cloning into '/tmp/buildpacks/java-buildpack'...
Note: checking out '26cf03bdf49c93f01793d0f2753e41a3e7061bc9'.
You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.
If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:
  git checkout -b new_branch_name
-----> Java Buildpack Version: v3.14 | https://github.com/rakutentech/java-buildpack#26cf03b
-----> Downloading Open Jdk JRE 1.8.0_121 from https://java-buildpack.cloudfoundry.org/openjdk/trusty/x86_64/openjdk-1.8.0_121.tar.gz (0.5s)
       Expanding Open Jdk JRE to .java-buildpack/open_jdk_jre (1.0s)
-----> Downloading Open JDK Like Memory Calculator 2.0.2_RELEASE from https://java-buildpack.cloudfoundry.org/memory-calculator/trusty/x86_64/memory-calculator-2.0.2_RELEASE.tar.gz (0.0s)
       Memory Settings: -XX:MaxMetaspaceSize=64M -Xss228K -Xms317161K -XX:MetaspaceSize=64M -Xmx317161K
-----> Downloading Container Certificate Trust Store 2.0.0_RELEASE from https://java-buildpack.cloudfoundry.org/container-certificate-trust-store/container-certificate-trust-store-2.0.0_RELEASE.jar (0.0s)
       Adding certificates to .java-buildpack/container_certificate_trust_store/truststore.jks (0.3s)
-----> Downloading Spring Auto Reconfiguration 1.10.0_RELEASE from https://java-buildpack.cloudfoundry.org/auto-reconfiguration/auto-reconfiguration-1.10.0_RELEASE.jar (0.0s)
-----> Uploading droplet (78M)

0 of 1 instances running, 1 starting
1 of 1 instances running

App started


OK

App v3.14-512M was started using this command `CALCULATED_MEMORY=$($PWD/.java-buildpack/open_jdk_jre/bin/java-buildpack-memory-calculator-2.0.2_RELEASE -memorySizes=metaspace:64m..,stack:228k.. -memoryWeights=heap:65,metaspace:10,native:15,stack:10 -memoryInitials=heap:100%,metaspace:100% -stackThreads=300 -totMemory=$MEMORY_LIMIT) && JAVA_OPTS="-Djava.io.tmpdir=$TMPDIR -XX:OnOutOfMemoryError=$PWD/.java-buildpack/open_jdk_jre/bin/killjava.sh $CALCULATED_MEMORY -Djavax.net.ssl.trustStore=$PWD/.java-buildpack/container_certificate_trust_store/truststore.jks -Djavax.net.ssl.trustStorePassword=java-buildpack-trust-store-password" && SERVER_PORT=$PORT eval exec $PWD/.java-buildpack/open_jdk_jre/bin/java $JAVA_OPTS -cp $PWD/. org.springframework.boot.loader.JarLauncher`

Showing health and status for app v3.14-512M
OK

requested state: started
instances: 1/1
usage: 512M x 1 instances
urls: api-jggkf.rpaas.net
last uploaded: Tue Mar 7 09:09:58 UTC 2017
buildpack: https://github.com/rakutentech/java-buildpack#v3.14

     state     since                    cpu      memory         disk           details
#0   running   2017-03-07 06:10:38 PM   376.2%   436M of 512M   158.2M of 1G
```

# a914fd6 behavior
```
Starting app master-128M
-----> Downloaded app package (32M)
Cloning into '/tmp/buildpacks/java-buildpack'...
-----> Java Buildpack Version: a914fd6 | https://github.com/rakutentech/java-buildpack#a914fd6
-----> Downloading Open Jdk JRE 1.8.0_121 from https://java-buildpack.cloudfoundry.org/openjdk/trusty/x86_64/openjdk-1.8.0_121.tar.gz (0.5s)
       Expanding Open Jdk JRE to .java-buildpack/open_jdk_jre (0.9s)
-----> Downloading Open JDK Like Memory Calculator 3.2.0_RELEASE from https://java-buildpack.cloudfoundry.org/memory-calculator/trusty/x86_64/memory-calculator-3.2.0_RELEASE.tar.gz (0.0s)
       Loaded Classes: 17882, Threads: 300, JAVA_OPTS: ''
-----> Downloading Container Certificate Trust Store 2.0.0_RELEASE from https://java-buildpack.cloudfoundry.org/container-certificate-trust-store/container-certificate-trust-store-2.0.0_RELEASE.jar (0.0s)
       Adding certificates to .java-buildpack/container_certificate_trust_store/truststore.jks (0.3s)
-----> Downloading Spring Auto Reconfiguration 1.10.0_RELEASE from https://java-buildpack.cloudfoundry.org/auto-reconfiguration/auto-reconfiguration-1.10.0_RELEASE.jar (0.1s)
-----> Uploading droplet (78M)

0 of 1 instances running, 1 down
0 of 1 instances running, 1 down
0 of 1 instances running, 1 down
0 of 1 instances running, 1 down
0 of 1 instances running, 1 down
0 of 1 instances running, 1 down
```
```
Starting app master-256M
-----> Downloaded app package (32M)
Cloning into '/tmp/buildpacks/java-buildpack'...
-----> Java Buildpack Version: a914fd6 | https://github.com/rakutentech/java-buildpack#a914fd6
-----> Downloading Open Jdk JRE 1.8.0_121 from https://java-buildpack.cloudfoundry.org/openjdk/trusty/x86_64/openjdk-1.8.0_121.tar.gz (0.5s)
       Expanding Open Jdk JRE to .java-buildpack/open_jdk_jre (1.0s)
-----> Downloading Open JDK Like Memory Calculator 3.2.0_RELEASE from https://java-buildpack.cloudfoundry.org/memory-calculator/trusty/x86_64/memory-calculator-3.2.0_RELEASE.tar.gz (0.0s)
       Loaded Classes: 17882, Threads: 300, JAVA_OPTS: ''
-----> Downloading Container Certificate Trust Store 2.0.0_RELEASE from https://java-buildpack.cloudfoundry.org/container-certificate-trust-store/container-certificate-trust-store-2.0.0_RELEASE.jar (0.0s)
       Adding certificates to .java-buildpack/container_certificate_trust_store/truststore.jks (0.3s)
-----> Downloading Spring Auto Reconfiguration 1.10.0_RELEASE from https://java-buildpack.cloudfoundry.org/auto-reconfiguration/auto-reconfiguration-1.10.0_RELEASE.jar (0.1s)
-----> Uploading droplet (78M)

0 of 1 instances running, 1 down
0 of 1 instances running, 1 down
0 of 1 instances running, 1 down
0 of 1 instances running, 1 down
0 of 1 instances running, 1 down
0 of 1 instances running, 1 down
```
```
Starting app master-512M
-----> Downloaded app package (32M)
Cloning into '/tmp/buildpacks/java-buildpack'...
-----> Java Buildpack Version: a914fd6 | https://github.com/rakutentech/java-buildpack#a914fd6
-----> Downloading Open Jdk JRE 1.8.0_121 from https://java-buildpack.cloudfoundry.org/openjdk/trusty/x86_64/openjdk-1.8.0_121.tar.gz (0.5s)
       Expanding Open Jdk JRE to .java-buildpack/open_jdk_jre (1.0s)
-----> Downloading Open JDK Like Memory Calculator 3.2.0_RELEASE from https://java-buildpack.cloudfoundry.org/memory-calculator/trusty/x86_64/memory-calculator-3.2.0_RELEASE.tar.gz (0.0s)
       Loaded Classes: 17882, Threads: 300, JAVA_OPTS: ''
-----> Downloading Container Certificate Trust Store 2.0.0_RELEASE from https://java-buildpack.cloudfoundry.org/container-certificate-trust-store/container-certificate-trust-store-2.0.0_RELEASE.jar (0.0s)
       Adding certificates to .java-buildpack/container_certificate_trust_store/truststore.jks (0.3s)
-----> Downloading Spring Auto Reconfiguration 1.10.0_RELEASE from https://java-buildpack.cloudfoundry.org/auto-reconfiguration/auto-reconfiguration-1.10.0_RELEASE.jar (0.0s)
-----> Uploading droplet (78M)

0 of 1 instances running, 1 down
0 of 1 instances running, 1 down
0 of 1 instances running, 1 down
0 of 1 instances running, 1 down
0 of 1 instances running, 1 down
0 of 1 instances running, 1 down
```

**Comment by nebhale** [comment]
At some point (I'm unclear on exactly when but @sclevine may have a precise date), the `$MEMORY_LIMIT` available during staging was disconnected from the `$MEMORY_LIMIT` used at runtime.  Therefore what used to work for us, running the memory calculation during staging and failing early if it wasn't going to work (as an early warning system) was no longer useful because [the value was always `1G`](https://github.com/cloudfoundry/cloud_controller_ng/blob/8b638405ac9c1eb36bd4d830d5bd1163cdaecae5/bosh/jobs/cloud_controller_ng/spec#L645) regardless of what the application's memory was tied to.  It was, sadly, removed part of the recent memory calculator changes.

I agree with you that the early calculation was always a help previously (even if `cf scale` meant that what passed during staging might not pass at runtime) but unfortunately, we can no longer do it.

**Comment by nebhale** [comment]
And now I realized that I may have completed missed helping you with the root cause.  If you look at the logs, you'll see the failure preventing your application from starting up.  Due to recent changes in the memory-calculator the effective minimum memory of a container is 768M.  We now account for a lot more memory spaces than we did with the v2 memory calculator.  You can see [our analysis here](https://docs.google.com/document/d/1vlXBiwRIjwiVcbvUGYMrxx2Aw1RVAtxq3iuZ3UK2vXA/edit#heading=h.uy41ishpv9zc), but the long and short is that Tomcat's 200 threads with 1M stacks, 240M for JIT Code Caching, a reasonable (> 100M) metaspace, already has you over 512M before you've even looked at heap.

**Comment by giner** [comment]
I see. Does this mean that we should always set 1024M limit as a minimum for a Java application if we do not override calculation manually? And if so wouldn't it make sense to add a warning somewhere so it would appear in logs?

**Comment by nebhale** [comment]
Indeed.  I think small applications can easily fit in 768M, but in reality the CF default of 1G containers is probably right and there shouldn't need to be any memory calculation customization from there.  If you find that the default values don't work for your application, please let me know; we're still tweaking the new memory calculator to cover nearly all applications and would welcome the input.

**Comment by giner** [comment]
@nebhale, I have deployed very simple "Hello world" application with logging to STDOUT enabled however it does not start with 512M (expected) and does not provide any meaningful logs:
```
$ cf logs hello-java-app --recent
Connected, dumping recent logs for app hello-java-app

2017-03-08T08:57:09.85+0900 [API/0]      OUT Created app with guid c1594e92-9a49-43a4-aee7-9026bdf2da4c
2017-03-08T08:57:10.32+0900 [API/0]      OUT Updated app with guid c1594e92-9a49-43a4-aee7-9026bdf2da4c ({"route"=>"666147d4-7f8e-49c6-8110-cc670ebaec23", :verb=>"add", :relation=>:routes, :related_guid=>"666147d4-7f8e-49c6-8110-cc670ebaec23"})
2017-03-08T08:57:16.76+0900 [DEA/1]      OUT Got staging request for app with id c1594e92-9a49-43a4-aee7-9026bdf2da4c
2017-03-08T08:57:16.77+0900 [API/0]      OUT Updated app with guid c1594e92-9a49-43a4-aee7-9026bdf2da4c ({"state"=>"STARTED"})
2017-03-08T08:57:18.62+0900 [STG/1]      OUT -----> Downloaded app package (26M)
2017-03-08T08:57:18.77+0900 [STG/0]      ERR Cloning into '/tmp/buildpacks/java-buildpack'...
2017-03-08T08:57:24.89+0900 [STG/0]      OUT -----> Java Buildpack Version: a914fd6 | https://github.com/rakutentech/java-buildpack#a914fd6
2017-03-08T08:57:26.99+0900 [STG/0]      OUT -----> Downloading Open Jdk JRE 1.8.0_121 from https://java-buildpack.cloudfoundry.org/openjdk/trusty/x86_64/openjdk-1.8.0_121.tar.gz (1.2s)
2017-03-08T08:57:28.00+0900 [STG/0]      OUT        Expanding Open Jdk JRE to .java-buildpack/open_jdk_jre (1.0s)
2017-03-08T08:57:28.10+0900 [STG/0]      OUT -----> Downloading Open JDK Like Memory Calculator 3.2.0_RELEASE from https://java-buildpack.cloudfoundry.org/memory-calculator/trusty/x86_64/memory-calculator-3.2.0_RELEASE.tar.gz (0.0s)
2017-03-08T08:57:28.53+0900 [STG/0]      OUT        Loaded Classes: 16492, Threads: 300, JAVA_OPTS: ''
2017-03-08T08:57:28.57+0900 [STG/0]      OUT -----> Downloading Container Certificate Trust Store 2.0.0_RELEASE from https://java-buildpack.cloudfoundry.org/container-certificate-trust-store/container-certificate-trust-store-2.0.0_RELEASE.jar (0.0s)
2017-03-08T08:57:28.90+0900 [STG/0]      OUT        Adding certificates to .java-buildpack/container_certificate_trust_store/truststore.jks (0.3s)
2017-03-08T08:57:29.13+0900 [STG/0]      OUT -----> Downloading Tomcat Instance 8.0.41 from https://java-buildpack.cloudfoundry.org/tomcat/tomcat-8.0.41.tar.gz (0.2s)
2017-03-08T08:57:29.24+0900 [STG/0]      OUT        Expanding Tomcat Instance to .java-buildpack/tomcat (0.1s)
2017-03-08T08:57:29.29+0900 [STG/0]      OUT -----> Downloading Tomcat Lifecycle Support 2.5.0_RELEASE from https://java-buildpack.cloudfoundry.org/tomcat-lifecycle-support/tomcat-lifecycle-support-2.5.0_RELEASE.jar (0.0s)
2017-03-08T08:57:29.33+0900 [STG/0]      OUT -----> Downloading Tomcat Logging Support 2.5.0_RELEASE from https://java-buildpack.cloudfoundry.org/tomcat-logging-support/tomcat-logging-support-2.5.0_RELEASE.jar (0.0s)
2017-03-08T08:57:29.38+0900 [STG/0]      OUT -----> Downloading Tomcat Access Logging Support 2.5.0_RELEASE from https://java-buildpack.cloudfoundry.org/tomcat-access-logging-support/tomcat-access-logging-support-2.5.0_RELEASE.jar (0.0s)
2017-03-08T08:57:38.82+0900 [STG/1]      OUT -----> Uploading droplet (78M)
2017-03-08T08:57:47.77+0900 [DEA/1]      OUT Starting app instance (index 0) with guid c1594e92-9a49-43a4-aee7-9026bdf2da4c
2017-03-08T08:57:50.66+0900 [DEA/1]      ERR Instance (index 0) failed to start accepting connections
2017-03-08T08:57:50.66+0900 [API/0]      OUT App instance exited with guid c1594e92-9a49-43a4-aee7-9026bdf2da4c payload: {"cc_partition"=>"default", "droplet"=>"c1594e92-9a49-43a4-aee7-9026bdf2da4c", "version"=>"8117b551-4e9b-459d-9734-9057acfc899e", "instance"=>"4c6b619fd3af4e2685fb912235813791", "index"=>0, "reason"=>"CRASHED", "exit_status"=>1, "exit_description"=>"failed to accept connections within health check timeout", "crash_timestamp"=>1488931070}
2017-03-08T08:57:50.66+0900 [API/0]      OUT App instance exited with guid c1594e92-9a49-43a4-aee7-9026bdf2da4c payload: {"cc_partition"=>"default", "droplet"=>"c1594e92-9a49-43a4-aee7-9026bdf2da4c", "version"=>"8117b551-4e9b-459d-9734-9057acfc899e", "instance"=>"4c6b619fd3af4e2685fb912235813791", "index"=>0, "reason"=>"CRASHED", "exit_status"=>1, "exit_description"=>"failed to accept connections within health check timeout", "crash_timestamp"=>1488931070}
2017-03-08T08:58:10.90+0900 [DEA/1]      OUT Starting app instance (index 0) with guid c1594e92-9a49-43a4-aee7-9026bdf2da4c
2017-03-08T08:58:13.53+0900 [DEA/1]      ERR Instance (index 0) failed to start accepting connections
2017-03-08T08:58:13.54+0900 [API/0]      OUT App instance exited with guid c1594e92-9a49-43a4-aee7-9026bdf2da4c payload: {"cc_partition"=>"default", "droplet"=>"c1594e92-9a49-43a4-aee7-9026bdf2da4c", "version"=>"8117b551-4e9b-459d-9734-9057acfc899e", "instance"=>"75e9c55ebf48470c9521ba4f0e4ff06d", "index"=>0, "reason"=>"CRASHED", "exit_status"=>1, "exit_description"=>"failed to accept connections within health check timeout", "crash_timestamp"=>1488931093}
2017-03-08T08:58:13.54+0900 [API/0]      OUT App instance exited with guid c1594e92-9a49-43a4-aee7-9026bdf2da4c payload: {"cc_partition"=>"default", "droplet"=>"c1594e92-9a49-43a4-aee7-9026bdf2da4c", "version"=>"8117b551-4e9b-459d-9734-9057acfc899e", "instance"=>"75e9c55ebf48470c9521ba4f0e4ff06d", "index"=>0, "reason"=>"CRASHED", "exit_status"=>1, "exit_description"=>"failed to accept connections within health check timeout", "crash_timestamp"=>1488931093}
```

**Comment by giner** [comment]
Is there any easy way to know the calculated settings so that we knew where to start adjustment from for our "Hello world"? In past we had the following message in logs `Memory Settings: -Xss228K -Xmx104169K -XX:MaxMetaspaceSize=64M -Xms104169K -XX:MetaspaceSize=64M` but it is not there anymore.

**Comment by nebhale** [comment]
Unfortunately, those logs were inaccurate as of some point recently.  Regardless of the amount of memory, they were always calculated against 1G.  A running VM would tell you what they were, but you can't get a running VM.  I'll need to see what we can do to print those out as part of the calculation at runtime (unfortunately we can't use `stdout` to do so) but it isn't there at the moment.

Your application should start with 768M (that's what we're seeing as an effective minimum with the default JVM and Tomcat values), but I have no idea why you're not seeing *any* logs at all in your console.  You'd expect to see some sort of OOM error and a stack trace typically (that's what we see when we start with 512M) but something seems to be suppresssing that.

**Comment by nebhale** [comment]
This is what we see on PWS for example when running a WAR with only 512M:

```
2017-03-08T06:41:15.16-0800 [STG/0]      OUT -----> Java Buildpack Version: a914fd6 | https://github.com/cloudfoundry/java-buildpack.git#a914fd6
2017-03-08T06:41:15.45-0800 [STG/0]      OUT -----> Downloading Open Jdk JRE 1.8.0_121 from https://java-buildpack.cloudfoundry.org/openjdk/trusty/x86_64/openjdk-1.8.0_121.tar.gz (found in cache)
2017-03-08T06:41:16.61-0800 [STG/0]      OUT        Expanding Open Jdk JRE to .java-buildpack/open_jdk_jre (1.1s)
2017-03-08T06:41:16.64-0800 [STG/0]      OUT -----> Downloading Open JDK Like Memory Calculator 3.2.0_RELEASE from https://java-buildpack.cloudfoundry.org/memory-calculator/trusty/x86_64/memory-calculator-3.2.0_RELEASE.tar.gz (0.0s)
2017-03-08T06:41:17.34-0800 [STG/0]      OUT        Loaded Classes: 13821, Threads: 300, JAVA_OPTS: '-XX:MaxMetaspaceSize=250M'
2017-03-08T06:41:17.37-0800 [STG/0]      OUT -----> Downloading Container Certificate Trust Store 2.0.0_RELEASE from https://java-buildpack.cloudfoundry.org/container-certificate-trust-store/container-certificate-trust-store-2.0.0_RELEASE.jar (0.0s)
2017-03-08T06:41:17.95-0800 [STG/0]      OUT        Adding certificates to .java-buildpack/container_certificate_trust_store/truststore.jks (0.5s)
2017-03-08T06:41:17.97-0800 [STG/0]      OUT -----> Downloading Container Customizer 1.1.0_RELEASE from https://java-buildpack.cloudfoundry.org/container-customizer/container-customizer-1.1.0_RELEASE.jar (found in cache)
2017-03-08T06:41:17.99-0800 [STG/0]      OUT -----> Downloading Spring Auto Reconfiguration 1.10.0_RELEASE from https://java-buildpack.cloudfoundry.org/auto-reconfiguration/auto-reconfiguration-1.10.0_RELEASE.jar (found in cache)
2017-03-08T06:41:18.04-0800 [STG/0]      OUT -----> Downloading YourKit Profiler 2017.02.53 from https://java-buildpack.cloudfoundry.org/your-kit/trusty/x86_64/your-kit-2017.02.53.so (found in cache)
2017-03-08T06:41:18.19-0800 [STG/0]      OUT -----> Downloading Tomcat Instance 8.0.41 from https://java-buildpack.cloudfoundry.org/tomcat/tomcat-8.0.41.tar.gz (0.1s)
2017-03-08T06:41:18.35-0800 [STG/0]      OUT        Expanding Tomcat Instance to .java-buildpack/tomcat (0.1s)
2017-03-08T06:41:18.38-0800 [STG/0]      OUT -----> Downloading Tomcat Lifecycle Support 2.5.0_RELEASE from https://java-buildpack.cloudfoundry.org/tomcat-lifecycle-support/tomcat-lifecycle-support-2.5.0_RELEASE.jar (0.0s)
2017-03-08T06:41:18.42-0800 [STG/0]      OUT -----> Downloading Tomcat Logging Support 2.5.0_RELEASE from https://java-buildpack.cloudfoundry.org/tomcat-logging-support/tomcat-logging-support-2.5.0_RELEASE.jar (0.0s)
2017-03-08T06:41:18.45-0800 [STG/0]      OUT -----> Downloading Tomcat Access Logging Support 2.5.0_RELEASE from https://java-buildpack.cloudfoundry.org/tomcat-access-logging-support/tomcat-access-logging-support-2.5.0_RELEASE.jar (0.0s)
2017-03-08T06:41:23.20-0800 [CELL/0]     OUT Destroying container
2017-03-08T06:41:24.63-0800 [CELL/0]     OUT Successfully destroyed container
2017-03-08T06:41:37.76-0800 [STG/0]      OUT Exit status 0
2017-03-08T06:41:37.76-0800 [STG/0]      OUT Uploading droplet, build artifacts cache...
2017-03-08T06:41:37.76-0800 [STG/0]      OUT Uploading build artifacts cache...
2017-03-08T06:41:37.76-0800 [STG/0]      OUT Uploading droplet...
2017-03-08T06:41:41.45-0800 [STG/0]      OUT Uploaded build artifacts cache (187.6M)
2017-03-08T06:41:45.14-0800 [STG/0]      OUT Uploaded droplet (69.6M)
2017-03-08T06:41:45.15-0800 [STG/0]      OUT Uploading complete
2017-03-08T06:41:45.18-0800 [STG/0]      OUT Destroying container
2017-03-08T06:41:45.85-0800 [CELL/0]     OUT Creating container
2017-03-08T06:41:46.53-0800 [CELL/0]     OUT Successfully created container
2017-03-08T06:41:49.79-0800 [STG/0]      OUT Successfully destroyed container
2017-03-08T06:41:50.10-0800 [CELL/0]     OUT Starting health monitoring of container
2017-03-08T06:41:50.22-0800 [APP/PROC/WEB/0]OUT Exit status 1
2017-03-08T06:41:50.22-0800 [APP/PROC/WEB/0]ERR Cannot calculate memory: insufficient memory remaining for heap. Memory limit 512M < allocated memory 837613K (-XX:ReservedCodeCacheSize=240M, -XX:MaxDirectMemorySize=10M, -XX:MaxMetaspaceSize=250M, -XX:CompressedClassSpaceSize=18413K, -Xss1M * 300 threads)
2017-03-08T06:41:50.24-0800 [CELL/0]     OUT Exit status 0
2017-03-08T06:41:50.26-0800 [CELL/0]     OUT Destroying container
2017-03-08T06:41:50.26-0800 [API/2]      OUT Process has crashed with type: "web"
2017-03-08T06:41:50.28-0800 [API/2]      OUT App instance exited with guid 820bf39e-2d1b-42b2-992e-4066c7f6356e payload: {"instance"=>"", "index"=>0, "reason"=>"CRASHED", "exit_description"=>"2 error(s) occurred:\n\n* 2 error(s) occurred:\n\n* Exited with status 1\n* cancelled\n* cancelled", "crash_count"=>1, "crash_timestamp"=>1488984110250649647, "version"=>"9086170e-c51a-4783-8315-79287eb75ad7"}
2017-03-08T0
```

**Comment by giner** [comment]
The example above is for Diego. Can it be related that we run on DEA?

**Comment by nebhale** [comment]
The lack of output absolutely could be. I'm not sure what our commitments are to support DEAs at this point, but it's likely that the 3.x line (the current changes in `master` are for a 4.x release) will be the place for DEA support.  Let me check into that today.

**Comment by nebhale** [comment]
I've confirmed this morning that the problem is the DEAs.  DEAs will be EOL in Cloud Foundry in mid-May or so, so we won't be supporting them with the code in `master` (we don't even have any test environments running DEAs left).  So, the recommendation for you, as long as you're on a DEA-based system is to use `https://github.com/cloudfoundry/java-buildpack.git#3.x` or `https://github.com/cloudfoundry/java-buildpack.git#v3.14` for compatible versions.

**Comment by CAFxX** [comment]
@nebhale doesn't that mean that DEA users of the java buildpack will be forced to run on unsupported versions of the buildpack for 2-3 months? Wouldn't it be better to wait until DEA EoL to actually drop compatibility and support? What's the point of announcing a date for DEA EoL if then in practice it reaches EoL before because of lack of updates?

**Comment by CAFxX** [comment]
>  You can see our analysis here, but the long and short is that Tomcat's 200 threads with 1M stacks, 240M for JIT Code Caching, a reasonable (> 100M) metaspace, already has you over 512M before you've even looked at heap.

(I'm assuming you meant 300 threads) While it's true that given the requirements above your JVM can't run with 512M, I don't agree with the requirements themselves - and the analysis you linked doesn't provide any rationale for e.g. the number of threads. Why 300 threads? Not all java apps spin up that many threads. If anything, I'd say that the maximum number of threads that can be supported should logically be proportional to the available memory (**since** they use memory). If you target 300 threads for 1G, then 512M should definitely not allow more than 150 - and if you factor in the "incompressible" 240M code cache, probably no more than 100. 
Other example: the code cache size of 240M - this is the default but it doesn't have to be fixed (it can be changed) and there may be good reasons to do so if the documentation is correct (https://docs.oracle.com/javase/8/embedded/develop-apps-platforms/codecache.htm#A1100181 "For most Java applications, this size is so large that the application will never fill the entire codecache") and this is before even considering that the code cache can be flushed if it's full.

I'm no Java expert but it definitely looks like there is space for some tradeoff to accommodate instances with smaller memory quotas.

**Comment by nebhale** [comment]
@CAFxX The 3.x line (with DEA support) will continue for some time in parallel with the 4.x line.  This lifespan will exceed the DEA EOL date and ensure that common updates will be available for people who are planning on sticking with the 3.x line for a while.

**Comment by CAFxX** [comment]
@nebhale wouldn't supporting DEAs involve testing the buildpack on DEAs? :sob:

**Comment by nebhale** [comment]
>  Not all java apps spin up that many threads.

Correct, and that number can be [configured in the buildpack](https://github.com/cloudfoundry/java-buildpack/blob/master/config/open_jdk_jre.yml#L28) to be accurate to what you're expecting if our default isn't correct.  We've picked 200 because Tomcat is far and away the most popular container in Cloud Foundry, and it [defaults to having a maximum of 200 threads in connectors](http://tomcat.apache.org/tomcat-8.5-doc/config/http.html).  And given that an idling VM has another ~20 threads before those 200 connections have been made, 

![screen shot 2017-03-08 at 8 37 18 pm](https://cloud.githubusercontent.com/assets/60754/23736166/3993beca-043f-11e7-820d-6beacb932cd4.png)

plus a bit extra for the possibility of thread pools in the application, we took a stab at 300.  That's subject to change of course, but is our starting point.  The memory calculator needs to handle the possibility of the maximum number of threads possible otherwise, the the JVM may request too much memory under heavy load and be terminated by the container at the point where it will do the most harm.  Given that the buildpack has no easy way to influence the number of threads in the container (especially as embedded Tomcat becomes the predominant way to run applications), we make a guess and allow the user to override if it's not appropriate.

> this is the default but it doesn't have to be fixed (it can be changed) and there may be good reasons to do so

Like the number of threads, the code cache size is configurable as well.  You can `cf set-env <APP> JAVA_OPTS '-XX:ReservedCodeCacheSize=<VALUE>'` and the memory calculator will honor it.  As you can see in our analysis document, our initial plan was to strongly constrain the code cache size (24M even in the example), but once we started seeing [reports of performance issues related to code cache size](http://engineering.indeedblog.com/blog/2016/09/job-search-web-app-java-8-migration/) we decided to do some performance testing.  In our experimentation we were able to see non-trivial CPU usage decreases to the 240M default and even beyond.  Rather than attempt a heuristic on something as performance sensitive as this, we decided to stick with the JVM default as this was something well known and well understood at this point; people who were comfortable tuning it could, and everyone else would get the same behavior as every other JVM in the world.

---
The new version of the memory calculator effort exists to address two major complaints:

1. The memory calculator doesn't actually calculate memory properly leading to situations where the container terminates JVMs with an inscrutable `exit 137`.  This is because it both doesn't take into account all memory spaces allocated by the JVM (code cache, compressed class space) and doesn't bound certain JVM spaces (metaspace, direct memory), leaving them to grow uncontrolled beyond the container's limits
1. When the memory calculator's choices were incorrect, overriding those choices was very hard and not very easy to reason about (`memory_heuristics`).

We did the work with mantra that the container should _never_ be allowed to terminate a process; a JVM's OoM error (for the various types of memory) was always more informative than `exit 137`.  To achieve this, we bound every memory space that we identified; if we missed some or new ones are created, we'll bound those too.  Where we think we safely can we calculate a value (metaspace) we will, but in places where our choices have measurable effects on performance (code cache), we'll use the defaults provided by the JVM.  In **all** cases, overriding our choices requires the standard JVM arguments used to configure those memory spaces.

**Comment by CAFxX** [comment]
Let me just take a step back for context. 

The problem we are reporting is a **regression**: applications that used to run with 512M (and even less) are not running anymore and this is causing quite some confusion to our users. I absolutely commend the effort to improve the error messages when memory is not enough, but this is undeniably causing a regression.

Resolving the regression would mean that a java app can start again when assigned a 512M quota.

With that clear:

> that number can be configured in the buildpack

configuring the buildpack like this seem to involve a fork with the overhead this requires; any chance to have this configurable via environment variable?

> plus a bit extra **for the possibility of thread pools in the application**, we took a stab at 300. That's subject to change of course, but is our starting point. **The memory calculator needs to handle the possibility of the maximum number of threads possible**

It seems to me the second phrase contradicts the first one. If there are thread pools in the application all bets are off and the `300` is as meaningless as `42` or `500`.

> We've picked 200 because Tomcat is far and away the most popular container in Cloud Foundry

While this is likely true, 200 is still the maximum number, and as mentioned above as soon as you even consider that an application may start threads outside of the connector pool, is not very meaningful.

This is a -admittedly very small- sample, but it's the best I can provide you in terms of actual numbers from one of our deployments (second column number of running threads, first column number of instances with that number of running threads):
```
      1 22
      1 24
      1 25
     11 26
     15 27
      6 28
     30 29
      4 30
      5 31
      2 32
      2 33
      3 34
      5 35
     22 36
      7 37
     12 38
      4 39
      7 40
      2 42
      4 43
      5 49
      1 52
      2 54
      1 55
      5 58
      3 59
      2 60
      2 61
      2 66
      2 67
      4 68
      2 69
      1 70
      1 71
      1 83
```

So, in our experience, the actual number of threads in use is significantly lower than the chosen number of 300. Maybe other operators can share their results to inform further tuning of this number (what if it turns out that 99% of apps use less than 100 threads, with the remaining 1% spread between 100 and 1000... do we set 1000 as the number of threads for calculation? or do we tune for the common case and clearly document how to tune the inform the memory calculator of any unusual application requirements?)

> As you can see in our analysis document, our initial plan was to strongly constrain the code cache size (24M even in the example), but once we started seeing reports of performance issues related to code cache size we decided to do some performance testing. 

I'm not advocating that the default should be lowered at the 1G point, I'm advocating that higher response times are better than the app not running at all at 512M. Maybe we could disable tiered compilation when running with less than 1G - this should shrink the default code cache size from 240 to 48M - at the likely (small-ish) expense of peak throughput.

**Comment by nebhale** [comment]
> but this effort is undeniably causing a regression.

In fact, not only do I believe that this is not a regression, I believe that it is actually a bug fix (although one that will definitely require changes to some applications).  While your applications have _started_ in 512M they cannot run under maximum load in 512M.  We've proven over and over, with dozens of customers that simply connecting 200 concurrent users (we use `siege` to replicate this) will cause applications to request more memory than the container will allow resulting in the instance being recycled.

> So, in our experience, the actual number of threads in use is significantly lower than the chosen number of 300.

The thread counts you sent are surely typical for your applications, but the memory calculator cannot calculate for the typical number of threads knowing that the container _could_ allow itself into a situation where it asks for too much memory.  The entire goal here to account for the worst possible case and ensure that the container will not recycle your application when it is most vulnerable.

If you find that your thread usage is significantly lower, I'd change the inputs to the memory calculator with an environment variable (`cf set-env <APP> JBP_CONFIG_OPEN_JDK_JRE '{ memory_calculator: { stack_threads: 1000 } }'`)  or a fork **as well as** changing your Tomcat configuration to ensure that it will never request more threads than this.  If you omit the latter, the application is vulnerable to an effective denial of service where high load leads to constant restarts.  This style of configuration [is documented](https://github.com/cloudfoundry/java-buildpack#configuration-and-extension) (although it could [use an update](https://github.com/cloudfoundry/java-buildpack/issues/393)).

> Maybe we could disable tiered compilation when running with less than 1G - this should shrink the default code cache size from 240 to 48M - at the likely (small-ish) expense of peak throughput.

As disabling tiered compilation _might_ have a performance impact, the buildpack won't do that, opting instead for the predictable JVM defaults around all performance-sensitive configuration.  If you feel safe that your applications will run acceptably without it they can opt into that behavior with, `cf set-env <APP> JAVA_OPTS '-XX:-TieredCompilation -XX:ReservedCodeCacheSize=48M'`.  Again this follows the policy of allowing users to judge the performance implications on an application-by-application basis and opt into changes using standard JVM flags.  To be clear, it's not that we believe that these changes are necessarily bad, we believe that they are not _predictable_ for users.

**Comment by TimGerlach** [comment]
After looking at some of the changes the memory calculator introduces with v3, I think there is another way to deploy applications with less than 768M.

The former memory calculator (2.0.2_RELEASE + [open_jdk_jre.yml](https://github.com/cloudfoundry/java-buildpack/blob/6f25b7e38a09188c9aeb2ff4f26dbbbaa8fdde5e/config/open_jdk_jre.yml) of java buildpack v3.x) calculated a stack size of 

- 349k for apps with 1G memory, 
- 262k for apps with 768M memory, and 
- 228k for apps with 500M memory 

while the new calculation algorithm uses 1M by default. Setting the stack size back to a smaller value (that used to work) helps to save a considerable amount of memory and finally succeed in calculating settings at all. Setting it back to 349k, for example, allowed me to successfully deploy my app with 500M memory again:

```shell
cf set-env my-application JAVA_OPTS: '-Xss349k'
```

This saves quite some memory, since:

349k x 300 ~ **102M**
1024k x 300 ~ **300M**

Also, this doesn't change the default Code Cache size considered by the calculator.

**Comment by nebhale** [comment]
Correct all around.  Previously the memory calculator tried to proportionally assign memory to different regions based on weightings as an attempt to optimize for the maximum amount of heap.  This caused numerous issues, and was considered incredibly difficult to adjust (you had to understand all of the proportional math to get the end result that you wanted).  This time around, we're assigning all the memory directly (defaults, straight multiplication of threads and stacks, and override with standard memory arguments) and whatever is left is used as the heap.  No longer does changing the container size change _all_ memory spaces, it simply adds all new space to the heap.

So if you believe that your application can run with smaller stacks than the default, you _shoud_ set `-Xss`; the memory calculator will take all the saved memory and move it to the heap.  If you believe that your application can run with a smaller code cache than the default, you _should_ set `-XX:ReservedCodeCacheSeize`; the memory calculator will take all the saved memory and move it to the heap.  This is by design and why we're sticking with the JVM defaults where they are already bounded (things like `metaspace` that are unbound, we do make a guess to ensure they can't grow forever, but it's overridable using the standard arguments).

**Comment by glyn** [comment]
@nebhale Thanks. This is something we should emphasise in the docs. Having a heading like "Tuning your application to work with smaller total memory" might be an idea.
